src/couch_mrview/src/couch_mrview_http.erl

```diff
@@ -709,3 +750,169 @@ maybe_set_page_size(#mrargs{page_size = undefined} = Args, MaxPageSize) ->
     Args#mrargs{page_size = MaxPageSize};
 maybe_set_page_size(#mrargs{} = Args, _MaxPageSize) ->
     Args.
+
+
+bookmark_encode(Args0) ->
+    {Args, _} = lists:foldl(fun(Value, {Acc, Idx}) ->
+        case element(Idx, Acc) of
+            Value ->
+                {setelement(Idx, Acc, []), Idx + 1};
+            _Default when Idx == #mrargs.bookmark
+                -> {setelement(Idx, Acc, []), Idx + 1};
+            _Default ->
+                {setelement(Idx, Acc, Value), Idx + 1}
+        end
+    end, {#mrargs{}, 1}, tuple_to_list(Args0)),
+    Term = erlang:delete_element(1, Args),
+    Binary = term_to_binary(Term, [compressed, {minor_version, 2}]),
+    couch_util:encodeBase64Url(Binary).
+
+
+bookmark_decode(Bookmark) ->
+    Binary = couch_util:decodeBase64Url(Bookmark),
+    Term = binary_to_term(Binary, [safe]),
+    {Args, _} = lists:foldl(fun(Value, {Acc, Idx}) ->
+        case {Value, element(Idx, Acc)} of
+            {[], []} ->
+                {Acc, Idx + 1};
+            {[], Default} ->
+                {setelement(Idx, Acc, Default), Idx + 1};
+            {_, _} when Idx == #mrargs.bookmark ->
+                {setelement(Idx, Acc, true), Idx + 1};
+            {Value, _} ->
+                {setelement(Idx, Acc, Value), Idx + 1}
+        end
+    end, {#mrargs{}, 1}, [mrargs | tuple_to_list(Term)]),
+    Args.
+
+
+paginated(Req, EtagTerm, #mrargs{page_size = PageSize} = Args, KeyFun, Fun) ->
+    Etag = couch_httpd:make_etag(EtagTerm),
+    chttpd:etag_respond(Req, Etag, fun() ->
+        hd(paginated(PageSize, [set_limit(Args)], KeyFun, Fun))
+    end).
+
+
+paginated(Req, EtagTerm, PageSize, QueriesArgs, KeyFun, Fun) when is_list(QueriesArgs) ->
+    Etag = couch_httpd:make_etag(EtagTerm),
+    chttpd:etag_respond(Req, Etag, fun() ->
+        Results = paginated(PageSize, QueriesArgs, KeyFun, Fun),
+        #{results => Results}
+    end).
+
+paginated(PageSize, QueriesArgs, KeyFun, Fun) when is_list(QueriesArgs) ->
```

Comment:
These three functions using the same name makes things surprisingly difficult to comprehend without having to stare at the arguments and work back why things would be different.